### ISSUE TITLE
undefined facts

### ISSUE THREAD

**Issue by calpa** [CacheControl/json-rules-engine#215]
When the user removed the input value, the value (fact) will becomes `undefined`.

Which throws an error according to the following code...

```js
if (typeof this.value === 'undefined' && typeof this.calculationMethod === 'undefined') {
  throw new Error('facts must have a value or method')
}
```

Link: https://github.com/CacheControl/json-rules-engine/blob/master/src/fact.js#L29

One possible solution is to pass `allowUndefinedFacts: true to the Fact Class`.

Related issues: https://github.com/CacheControl/json-rules-engine/issues/111

**Comment by saswat3115** [comment]
even if u pass ```allowUndefinedFacts: true``` this also fails.

**Comment by saswat3115** [comment]
Following changes need to be done to fix this

changes in fact.js
![image](https://user-images.githubusercontent.com/14299682/96709970-0d6c2200-13b9-11eb-9bd8-17a4282e8619.png)

changes in almanac.js
![image](https://user-images.githubusercontent.com/14299682/96710082-2c6ab400-13b9-11eb-8324-df12f63a663f.png)

**Comment by chenlevy** [comment]
Who's managing this rep? Can you please  implement the purposed solution

**Comment by CacheControl** [comment]
One point of clarity: `allowUndefinedFacts` controls the fact _identifier_, whereas the error being thrown above is the fact _value_:

```
{
  fact: 'my-fact',  <--- fact identifier, throws if undefined unless allowed via allowUndefinedFacts
  operator: 'equal',
  value: undefined, <---- fact value
}
```

I agree it's silly to require fact values to not be undefined - it just forces everyone to set their fact values to something else (`null`, `false`, etc). It also prevents valid use of `equal`/`notEqual` with `undefined`.

I've opened #236 to fix this, it is published as **5.1.0-rc1** I'll merge it in the next few days. @saswat3115 @calpa could you try installing this RC and validate it fixes your issue please?

**Comment by CacheControl** [comment]
This is now published as **5.3.0**

**Comment by gunjan-it-engg** [comment]
`const engine = new Engine(rules, { allowUndefinedFacts: true });`
